feature/todo/todo-ui/src/commonMain/kotlin/io/github/vasilyrylov/archsample/feature/todo/ui/screen/list/TodoListViewModel.kt

```diff
@@ -0,0 +1,72 @@
+package io.github.vasilyrylov.archsample.feature.todo.ui.screen.list
+
+import io.github.vasilyrylov.archsample.common.ui.base.BaseStateViewModel
+import io.github.vasilyrylov.archsample.common.ui.navigation.RouterHolder
+import io.github.vasilyrylov.archsample.common.domain.model.TodoItem
+import io.github.vasilyrylov.archsample.common.domain.model.TodoItemId
+import io.github.vasilyrylov.archsample.feature.todo.domain.api.ILogoutUseCase
+import io.github.vasilyrylov.archsample.feature.todo.domain.usecase.ObserveTodoListUseCase
+import io.github.vasilyrylov.archsample.feature.todo.domain.usecase.SaveTodoUseCase
+import io.github.vasilyrylov.archsample.feature.todo.domain.usecase.TodoCompletedChangeUseCase
+import io.github.vasilyrylov.archsample.feature.todo.ui.api.ITodoFlowRouter
+import io.github.vasilyrylov.archsample.feature.todo.ui.screen.list.model.TodoListViewState
+import io.github.vasilyrylov.archsample.feature.todo.ui.screen.list.model.TodoListScreenDialog
+import kotlinx.coroutines.launch
+
+class TodoListViewModel(
+    private val observeTodoList: ObserveTodoListUseCase,
+    private val routerHolder: RouterHolder<ITodoFlowRouter>,
+    private val completedChange: TodoCompletedChangeUseCase,
+    private val saveTodo: SaveTodoUseCase,
+    private val logout: ILogoutUseCase,
+) : BaseStateViewModel<TodoListViewState>() {
+
+    private val router: ITodoFlowRouter
+        get() = routerHolder.router
+
+    init {
+        initObservers()
+    }
+
+    fun onAddClick() {
+        setState { copy(dialog = TodoListScreenDialog.AddTodo) }
+    }
+
+    fun onConfirmAdd(newItem: TodoItem) {
+        coroutineScope.launch {
+            saveTodo(newItem)
+            setState { copy(dialog = TodoListScreenDialog.None) }
+        }
+    }
+
+    fun onCancelAdd() {
+        setState { copy(dialog = TodoListScreenDialog.None) }
+    }
+
+    fun onCompletedChange(id: TodoItemId) {
+        coroutineScope.launch {
+            completedChange(id = id)
+        }
+    }
+
+    fun onTodoClick(id: TodoItemId) {
+        router.toDetailTodo(id = id)
+    }
+
+    fun onLogoutClick() {
+        logout()
+    }
+
+    private fun initObservers() {
+        coroutineScope.launch {
```

Comment:
You can use [launchIn](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/launch-in.html) for this case